RFD 59: Search Based Access Requests (#11261)
gravitational/teleport: rfd/0059-search-based-access-requests.md

requests proposed here are useful for two different scenarios:

- Sometimes it is helpful to request elevated roles in the system, like
  `db-admin`. For example, Alice may need to run a Ansible playbook on all
  machines in a given class, or perform some system upgrades or troubleshooting.
  In this case, role-based access requests work great.

- In other cases, it’s helpful to request access to a subset of resources. In
  this case, search.

### Example Flow

Alice is a member of the group “splunk”.

She gets an alert that “db-1” is malfunctioning.

Alice needs to locate the "db-1" resource and request access to it.

Alice goes to the “Request access” screen and selects: “select access to resources”
option that is available alongside the “select access to roles” option.

Alice uses standard (fuzzy search) or advanced search (labels) to enter the
resource name.

She finds a single resource (e.g. "db-1") or a group of resources by that name and clicks
"request access".

An access request is created. It is sent to the "db-admins" team members via slack,
email or other configured channel based on the labels on the requested resources

> (we will add label-based routing to access request plugins).

Reviewer: I'd be interested in seeing a configuration example for the proposed routing of access requests to the request plugins.

Author: I added an example of the new configuration section here 812cc46

---
authors: Alexander Klizhentas (sasha@goteleport.com), Nic Klaassen (nic@goteleport.com)
state: draft
---

# RFD 59 - Search Based Access Requests

## What

This RFD proposes just-in-time search-based access requests which will enable
users to request access to resources based on a search and/or selection of
individual resources, rather than having to request one or more roles.

## Why

Teleport users may not know in advance which roles they need or who should grant
access, just which resource(s) they need access to.

Users would like to avoid having many different custom roles for different
levels of access.

Requesting access to a "blanket" role violates the principle of least priviledge.
Most of the time incident responders only need access to 1 or 2 nodes,
search-based access requests will allow them to request access to only the nodes

> they need.

Reviewer 2: I think this is well stated: there is a further point in here that a 'review' of a blanket access request is enormously harder than a specific request. 

if Alice has the role of splunk-admin, she may have access to tens of thousands of systems. Its extremely easy to generate a reason to login to one of them, but then connect to a different one, or all of them. Permitting access to a single box to troubleshoot (say) Splunk is very, very different to permitting somebody to run Ansible across the fleet. If we step back and consider that the concern is not just Alice suddenly turning malicious, but that an attacker could get her (short lived) secret, it makes no sense for Alice to have a secret that allows her to connect to all of prod to fix Splunk on a single one.

## Details

> Users will be able to request access by using search, and roles will be used

Reviewer 2: I think this is what you are proposing below, but its worth calling it out here: our preference is that users are not just able to request access by search - they are able to request access by simply connecting  to a host. The tsh infra then does the search for them and decides if they are in this new category of "can access, but with permission from <x slack room>". 

If they are in that group, they are then told that access to that resource requires a reason and permission, and the tsh ssh command just blocks waiting for that to be approved or denied. 

The user experience (i.e. tsh ssh <host> blocking) is much more important than how we wire up deciding who is in that group of 'sure, but <x> slack room needs to approve', since the former affects the usability for users (i.e. whether we can actually use this) whereas the latter is simply a question of what API we have to deal with (we can deal with any reasonable API).

Author: The original idea was for access requests based on search and that lead to the way this document was originally written, but automatically requesting access just by attempting to connect to a host is very powerful and may well end up being the more popular usecase. I will make sure it's a first-class feature here particular attention on the UX.

behind the scenes to evaluate who will be responsible for reviewing and granting
access requests.

Users won’t be asked to think about roles in advance, but will be able to focus
on finding and requesting access to individual resources.

The existing “role-based” access requests and the new “search-based” access
requests proposed here are useful for two different scenarios:

- Sometimes it is helpful to request elevated roles in the system, like
  `db-admin`, to perform some system upgrades, or system wide-troubleshooting.
  In this case, role-based access requests work great.

- In other cases, it’s helpful to request access to a subset of resources. In
  this case, search.

### Example Flow

Alice is a member of the group “splunk”.

She gets an alert that “db-1” is malfunctioning.

Alice needs to locate the "db-1" resource and request access to it.

> Alice goes to the “Request access” screen and selects: “select access to resources”

Reviewer 2: We would really like that the API is simply that Alice does 'tsh ssh db-1' and is told that she needs to give a reason and wait for permission from Slack group X. The vast majority of our users do not use the Teleport web UI (except as a clickable bounce to MFA auth) - they use their terminal. 

tsh can of course do this lookup via API, e.g. figure out that Alice is not a member of a role that has normal access to db-1, but she is in this group that can access it via permission. But from the users POV its all the same - some hosts they can get into, some hosts they cant, and there is this new middle ground. 

To be clear I think its fine for this to only work for 'tsh ssh' (rather than openssh), certainly initially.  If it can work in the web UI as well, thats great, but thats less important for us than the tsh command.

Author: We will make this work with `tsh ssh` as mentioned in other comments and in the doc now.

### CLI Mock-Up

```bash
$ tsh request search --kind db --labels 'env=prod'
Found 2 items:

name kind     id
db-1 database 388aff7f-459f-4a43-804a-3729854976ab
db-2 database 3be2fdad-7c79-4cfa-924e-ec1ea7225320

Create access request by:
> tsh request create --id 388aff7f-459f-4a43-804a-3729854976ab --id 3be2fdad-7c79-4cfa-924e-ec1ea7225320
```

Users can search by kind, labels, and keywords. The `tsh request search` command
will output a `tsh request create` command which can be used to request access.

The CLI UX may be improved by allowing users to perform multiple searches,
"stage" which resources they want to request access to and store this state
locally, and finally execute the full request without needing each UUID on the
command line. But this may be too complex and confusing. With this MVP version
users can copy-paste UUIDs from multiple searches, the meaning is clear, and it
is fully customizable and scriptable.

### On-Demand SSH

Many times users would not want to search and request access in two steps.
Teleport will have a new alias flag modifier that will try accessing a node, in
case if access denied, it will request access for a node all in one step:

> 

Reviewer 3: Async vs sync mode:

* Early version would not need `async` mode of requests. We might add those in the future if someone askss

How to deal with multiple nodes for `tsh ssh -P root@app=jenkins` for example?

* Look at the implementation complexity and decide whether to add to initial release or postpone. Consider limitations of how many nodes can be added to the request. 

Whether we want to support `-P` for databases and other resources

* Initial release should not focus on anything other than SSH.

```bash
# -P "Please" activate try and request mode
$ tsh ssh -P root@db-node
# by default, if tsh will succeed to SSH, it will just continue with a
# Session, otherwise it will find a node and create a search based access request
You do not have access to the system by default, created access request.

Please wait...

Access request has been approved. Happy hacking!
$
```

Some customers would want to have `-P` flag as a default. New alias feature in

> profile, will allow those users to set `tsh ssh` as an alias for `tsh ssh -P`.

Reviewer 2: This means what, a bashrc type alias? Or a teleport profile that we can set server side for all our users?

Author: The RFD for tsh aliases is here https://github.com/gravitational/teleport/pull/11447. But at this point I'm leaning toward the idea in the comments above of skipping the flag - if the user has any roles with `search_as_roles` specified which could grant access to the host, we should automatically prompt the user to create the access request.

For OpenSSH use-cases, some users would like to load the new certificates in the
agent, `tsh login -A root@db-node` will work like `tsh ssh -P` command except

> instead of login, it will load the keys in the agent.

Reviewer 2: This would obviously be great for some users, but this to me is a clear future/v2 feature.

with uuid “uuid-1” and “uuid-2”. In case if leaf’s uuid-1 has label “class:
external-access-allowed” and uuid-2 does not. Leaf cluster will reject access to
the node uuid-2 despite the fact that the root cluster “allowed” it.

### Access Requests RBAC Edge-Cases

A user could find and request access to several different types of resources,
some with `label: a` others with `label: b`.
Access to resource `label: a` can be granted by users having role `a`.
Access to resource `label: b` can be granted by users having role: `b`.
In this case both users with role `a` and role `b` have to approve access for
the request to be granted.

### Access Requests RBAC Plugin Notification

Access request slack and mattermost plugin should add routing to channels by
label and resource type.

Will complete this section with an example config, or create another RFD.

### Audit events

Will create a new audit event `access_request.search` that will include the searches users are
running in the scope of requesting access.

### OpenSSH

OpenSSH clients should work with credentials obtained via search-based access requests, assuming that the proxy and node are both Teleport.

> OpenSSH servers won't support this flow.

Reviewer 3: We can add `-P` flag to proxy command's `tsh proxy ssh -P` as a proxy command for OpenSSH to work

access requests.

Users won’t be asked to think about roles in advance, but will be able to focus
on finding and requesting access to individual resources.

The existing “role-based” access requests and the new “search-based” access
requests proposed here are useful for two different scenarios:

- Sometimes it is helpful to request elevated roles in the system, like
  `db-admin`, to perform some system upgrades, or system wide-troubleshooting.
  In this case, role-based access requests work great.

- In other cases, it’s helpful to request access to a subset of resources. In
  this case, search.

### Example Flow

Alice is a member of the group “splunk”.

She gets an alert that “db-1” is malfunctioning.

Alice needs to locate the "db-1" resource and request access to it.

Alice goes to the “Request access” screen and selects: “select access to resources”
option that is available alongside the “select access to roles” option.

Alice uses standard (fuzzy search) or advanced search (labels) to enter the
resource name.

She finds a single resource or a group of resources by that name and clicks
"request access".

Teleport identifies that access to the node “db-1” is managed by team

> “db-admins” based on Alice's role which allowed her to search for this node.

Reviewer 4: I'm confused, how does Alice's role that allowed her to search for this node enable teleport to identify that access to it is managed by "db-admins"?

Author: this sentence was wrong, the access request will be sent to the right channel based on the resource labels, we'll be updating the plugins to support this, clarified in the doc

An access request is created and sent to the "db-admins" team members via slack,
email or other configured channel.

There are two thresholds set in access requests for this group of resources.
When both Ivan and Mary approve the request, Alice is granted access for one
hour.

### Query Language and Filter UI

The query language and filter UI is defined by the
[Pagination and Search RFD](./0055-webui-ss-paginate-filter.md).

### CLI Mock-Up

```bash
$ tsh request search --kind db --labels 'env=prod'
Found 2 items:

name kind     id
db-1 database 388aff7f-459f-4a43-804a-3729854976ab
db-2 database 3be2fdad-7c79-4cfa-924e-ec1ea7225320

Create access request by:
> tsh request create --id 388aff7f-459f-4a43-804a-3729854976ab --id 3be2fdad-7c79-4cfa-924e-ec1ea7225320

> ```

Author: Yeah, should be pretty easy to add this flag

> ```

Reviewer 2: See above: we would much prefer that a user can just do "tsh ssh" and have it do what you are suggesting. What this means is that the users ahead of time dont need to know if they are in a role that can access db-1 without further permission, or not.

> ```

Author: mentioned this in the doc now

Users can search by kind, labels, and keywords. The `tsh request search` command
will output a `tsh request create` command which can be used to request access.

The CLI UX may be improved by allowing users to perform multiple searches,
"stage" which resources they want to request access to and store this state
locally, and finally execute the full request without needing each UUID on the
command line. But this may be too complex and confusing. With this MVP version
users can copy-paste UUIDs from multiple searches, the meaning is clear, and it
is fully customizable and scriptable.

> ### On-Demand SSH

Author: This also only gives access to 1 node at a time, so the suggestion is still relevant imo

Many times users would not want to search and request access in two steps.
Teleport will have a new alias flag modifier that will try accessing a node, in
case if access denied, it will request access for a node all in one step:

```bash
# -P "Please" activate try and request mode

> $ tsh ssh -P root@db-node

Author: Agreed, I just copied this from the google doc for now. This behaviour is interesting enough that it probably warrants a long flag, I would propose `--request`

> $ tsh ssh -P root@db-node

Reviewer 5: Having to ssh with a specific flag sort of defeats the purpose of this functionality though.  Additionally, if this workflow will support openssh can an informational message be generated for the user so they know their access is denied but an access request is generated and pending approval (though not sure whether openssh will support that).

> $ tsh ssh -P root@db-node

Reviewer 3: @jdconti see the section with aliases that will allow to have `tsh ssh` to work like `tsh ssh -P` for your org

> $ tsh ssh -P root@db-node

Reviewer 2: ^^ +1; if we can enable this by default that would be great. if we cant, can you have an env variable that enables it so we can (try) to set it globally for our users? AS noted above, its really ideal for us that our users dont have to remember to do anything different (different flag, different command, different workflow) to access this new type of hosts if they are using tsh ssh.

> $ tsh ssh -P root@db-node

Author: @jdconti @daviesalex @klizhentas I'm wary of changing the default behaviour globally and potentially causing people to accidentally create access requests, e.g. if they merely mistyped a hostname or something.

Would it be acceptable to have a prompt or a suggested command if the user cannot currently access the host but would be able to request access to it? Something like the following:
```bash
$ tsh ssh ubuntu@host1
ERROR: access denied to ubuntu connecting to host1 on cluster cluster-one

Request access (Y/n)? Y
Enter request reason: need to fix x
Creating request...
Request ID: c723ff91-536b-4125-b2bd-437aafa5ea15
Username:   ubuntu
Roles:      splunk
Resources:  388aff7f-459f-4a43-804a-3729854976ab (host1)
Reason:     "need to fix x"
Reviewers:  [none] (suggested)
Status:     PENDING

hint: use 'tsh login --request-id=<request-id>' to login with an approved request

Waiting for request approval...

Approval received, getting updated certificates...

ubuntu@host1:~$
```

> $ tsh ssh -P root@db-node

Reviewer 3: @daviesalex with aliases (see https://github.com/gravitational/teleport/pull/11447/files) you will be able to set this globally for your team

> $ tsh ssh -P root@db-node

Reviewer 3: @nklaassen the flow (Y/N) will defeat the purpose of `-R` that implies that is what users want (create a request if they can't access the node), so I think we don't need it

> $ tsh ssh -P root@db-node

Reviewer 2: @nklaassen thanks for your detail. Your workflow is I think perfect. 

Its not ideal that we would have to manage a file in every users environment, but I totally understand the desire to not confuse users [though would it be right that if users dont enable the new feature by not having any users in this new category of 'need to ask for permission for this host', the behaviour is not changing?]. In any case, I think we can deal with this. 

cc @jdconti for thoughts?

> $ tsh ssh -P root@db-node

Reviewer: One wrinkle I can see here is how to handle the availability of multiple roles that may be requested. If a user's role has multiple `search_as_roles` values which might apply to a host or the user has multiple roles each with `search_as_roles` values which might apply to the host, which are applied when the access request is granted?

> $ tsh ssh -P root@db-node

Reviewer 5: > Its not ideal that we would have to manage a file in every users environment

I agree, it would be helpful to support other methods of enabling automatic access request creation (e.g. environment, existing role, and perhaps user traits/metadata/claims).  While dotfiles are great for opinionated users managing them at scale can be painful; supporting a global `/etc/tshconfig` would ease that pain slightly (as I mentioned in #11447)

> $ tsh ssh -P root@db-node

Author: > One wrinkle I can see here is how to handle the availability of multiple roles that may be requested. If a user's role has multiple search_as_roles values which might apply to a host or the user has multiple roles each with search_as_roles values which might apply to the host, which are applied when the access request is granted?

If the user has multiple `search_as_roles` which could grant access to the requested resource, all such roles would be requested by default (with access limited to the specific resource requested). If for some reason a specific role is needed (which the user would know about in this case) the web UI and the `tsh request create` command will provide the option to request a specific role with a search-based access request.

> $ tsh ssh -P root@db-node

Reviewer 3: @nklaassen to follow the principle of least privilege, we should request access to the minimal role set that grants the privileges, for example  if Alice is executing `ssh alice@host`, and there are two roles: `a` -> grants access to `host` as `alice` and `b` that grants access as `root, alice`, only `a` should be requested.

> $ tsh ssh -P root@db-node

Author: @klizhentas it should be pretty straightforward to do this for `tsh ssh --request user@node` requests where we know the login, we can just choose a role with the minimal set of logins. When using the actual search flow we could try to do something similar by making the user put in the set of logins they need, but I think the simplest solution for that case is still to request all available search_as_roles by default, the cert will be limited to only the requested resources.

To enable this level of matching, we propose an extension to the role spec to
determine the final roles that need to be granted.

We will specify two roles that define this flow.

The role `response-team` will allow user with this role to
request access to nodes of teams “dbs” and “splunk”, and members of OIDC/SAML
group “app” developers to request access to their own group.

The spec for role-local variables will be expanded in a separate RFD (will post
link here).

```yaml
kind: role
metadata:
  name: response-team
spec:
  # Local variables are necessary to specify additional rules
  # that are not available in OIDC/SAML claims.
  # Teleport admins don't always have control over OIDC/SAML.
  # Specifying local variables allows them to add additional cases used in roles.
  # Having role-local variables simplifies computation and role evaluation,
  # as role eval keeps everything in one place.
  vars:
    - name: access_group
      select: external.claims[groups]:
        - "splunk": ["dbs", "{{external.claims['group']}}"]
        # by default, allow members of the group to request access to their own group
        - ".*": "{{external.claims['group']}}"
  allow:
    request:

>       # search_as_roles will allow a member of the response team
>       # to search resources that they will be allowed to request
>       # access to as a member of db-admin-role

Reviewer 4: ```suggestion
      # search_as_roles allows a member of the response team
      # to search for resources accessible to users with the db-admin-role,
      # which they will be allowed to request access to. If the access request
      # request is approved, the response team member will assume a restricted
      # version of the db-admin-role, which only has access to the specific resources
      # they were approved for.
```

      # delegate details of the existing permissions to a special
      # role that specifies that access.

>       # E.g the role can have only one logins: root parameter and that's all.

Reviewer 4: These clauses both confused me. Perhaps they can be removed, or at least clarified.

      search_as_roles: [db-admin-role]
```

Users will assume the role db-admin-role when searching for nodes in the search

> UI and will also receive the certificate with this role if access request is

Reviewer 4: ```suggestion
UI and will also receive the certificate with this role (but restricted to the specific resources they requested) if access request is
```

UI and will also receive the certificate with this role if access request is
granted.

```yaml
kind: role
metadata:
  name: db-admin-role
spec:
  # This role accepts a parameter "access_group". This parameter
  # is either supplied when evaluating the role during access check
  # and it should be present in the cert, or it is passed as a local
  # variable as a part of "search-as"
  params:

>    - name: access_group

Reviewer 6: Similar to my previous comment, if possible I think it would be nice for the first example to be as simple as possible, where the roles are already configured correctly and don't need additional parameterization.

    logins: ["root"]
    # db_labels defines what databases this role will be allowed to search
    # as a part of search_as request and in addition to that will be used to
    # evaluate access.
    db_labels:
       owner: "{{params.access_group}}"
```

### Certificate issuance and RBAC

Once a user assumes “db-admin-role” the Teleport root cluster issues a cert
including the assumed role and a list of UUIDs of resources granted by the
access request.

This makes sure that the role is scoped to a static list of resources that never
changes for this certificate.

```

> Assumed-role: [db-admin]

Reviewer 4: ```suggestion
Assumed-role: [db-admin-role]
```

Resource-UUIDs: [uuid-1, uuid-2]
```

### Trusted clusters

Leaf clusters will use standard role mapping to validate the cert issued by the

> root. If the leaf cluster role maps root cluster’s “db-admin” to the same role

Reviewer 4: ```suggestion
root. If the leaf cluster role maps root cluster’s “db-admin-role” to the same role
```

```yaml
role_map:
   '*': '*'
```

Then leaf cluster behavior will be identical. Leaf cluster may choose to narrow
the scope, for example:

```yaml
role_map:
   'db-admin-role': 'leaf-db-admin-role'
```

```yaml
kind: role
metadata:
  name: leaf-db-admin-role
spec:
  # This role accepts a parameter "access_group". This parameter
  # is either supplied when evaluating the role during access check
  # and it should be present in the cert, or it is passed as a local
  # variable as a part of "search-as"
  params:
   - name: access_group
  allow:
    logins: ["root"]
    # node_labels defines what nodes this role will be allowed to search
    # as a part of search_as request and in addition to that will be used to
evaluate access
    node_labels:
       owner: "{{params.access_group}}"
       class: external-access-allowed
```

In this case, if the cert issued by the root cluster granted access to nodes
with uuid “uuid-1” and “uuid-2”. In case if leaf’s uuid-1 has label “class:
external-access-allowed” and uuid-2 does not. Leaf cluster will reject access to
the node uuid-2 despite the fact that the root cluster “allowed” it.

### Access Requests RBAC Edge-Cases

> A user could find several different types of resources `label: a` and `label: b`.

Reviewer 6: ```suggestion
A user could find and request access to several different types of resources, some with `label: a` others with `label: b`.
```

Access to resource `label: a` can be granted by users having role `a`. Access
to resource `label: b` can be granted by users having role: `b`. In this case both
users with role `a` and role `b` have to approve access for the request to be
granted.

### Access Requests RBAC Plugin Notification

Access request slack and mattermost plugin should add routing to channels by
label and resource type.

Will complete this section with an example config, or create another RFD.

> ### Search events

Reviewer 6: ```suggestion
### Audit events
```

> Will create an event `access_request.search` that will include the searches users are

Reviewer 6: ```suggestion
Will create a new audit event `access_request.search` that will include the searches users are
```

running in the scope of requesting access.

### OpenSSH

> OpenSSH clients should work, assuming that the proxy and node are both Teleport.

Reviewer 6: ```suggestion
OpenSSH clients should work with credentials obtained via search-based access requests, assuming that the proxy and node are both Teleport.
```

There are a few options for this:

1. The
   [alias feature](https://github.com/gravitational/teleport/blob/master/rfd/0061-tsh-aliases.md)
   in the tsh profile will allow those users to set `tsh ssh as an alias for `tsh ssh --request`.
2. `tsh` can output a suggested command with the `--request` flag if access is
   denied but could be requested
3. `tsh` could prompt the user to create a request with a reason.

### Which roles will be requested

When creating a search-based access request the underlying roles being requested
will be determined automaticially. For simplicity, all roles which the user has
permission to search as (included in `search_as_roles` on any of the roles the
user has) will be requested. This request will be limited to only the exact
resources found in the search, and (if approved) the user will have access to
all logins granted by those roles.

For "on-demand ssh" (`tsh ssh --request user@node`) we will attempt to find and
request a single role which grants access to the node with the requested login.
If multiple such roles exist, the role with the lowest number of allowed logins
will be requested. In case of a tie, the requested role will be chosen

> arbitrarily.

Author: The difference is that for the first case ("search-based" request) we don't know which login the user needs, we delegate that to the role definition as mentioned in my other reply. For the "on-demand ssh" `tsh ssh user@node` it is specific to ssh-only and we know the login so we can try to be a bit more smart about how we pick the role.

Ideally both will be the same in practice if the user doesn't have multiple `search_as_roles` granting access to overlapping sets of nodes

### Certificate issuance and RBAC

Once a user assumes the “db-admins” role the Teleport root cluster issues a cert
including the assumed role and a list of UUIDs of resources granted by the
access request.

This makes sure that the role is scoped to a static list of resources that never
changes for this certificate.

```
Assumed-role: [db-admins]

> Resource-UUIDs: [node:uuid-1, node:uuid-2]

Author: Your understanding is correct, we are granting resource-level resolution but leaving the rest up to the role definition. We have many other restrictions defined in the role spec beyond the SSH principal, the current plan is to delegate all that configuration to the role and keep this feature somewhat minimal to start with.
